zerver/lib/actions.py

```diff
@@ -3569,27 +3575,54 @@ def get_non_subscribed_subs() -> List[Tuple[UserProfile, Stream]]:
     for stream_id in stream_admin_dict.keys():
         stream_admin_count_dict[stream_id] = len(stream_admin_dict[stream_id])
 
+    removed_stream_admin_list: List[SubInfo] = []
     for sub in subs_to_deactivate:
         user_profile_id = sub.user.id
         stream_id = sub.stream.id
         if user_profile_id in stream_admin_dict[stream_id]:
             if stream_admin_count_dict[stream_id] == 1:
                 raise JsonableError(_("Cannot remove all stream administrators."))
             stream_admin_count_dict[stream_id] -= 1
+            removed_stream_admin_list.append(sub)
 
     our_realm = users[0].realm
 
     # We do all the database changes in a transaction to ensure
     # RealmAuditLog entries are atomically created when making changes.
     with transaction.atomic():
         occupied_streams_before = list(get_occupied_streams(our_realm))
+        Subscription.objects.filter(
+            id__in=sub_ids_to_deactivate,
+            role=Subscription.ROLE_STREAM_ADMINISTRATOR,
+        ).update(role=Subscription.ROLE_MEMBER)
+
         Subscription.objects.filter(
             id__in=sub_ids_to_deactivate,
         ).update(active=False)
         occupied_streams_after = list(get_occupied_streams(our_realm))
 
         # Log Subscription Activities in RealmAuditLog
         event_time = timezone_now()
+
+        subscription_role_changed_logs = [
+            RealmAuditLog(
+                realm=sub.user.realm,
+                event_type=RealmAuditLog.SUBSCRIPTION_ROLE_CHANGED,
+                event_time=event_time,
+                modified_user=sub.user,
+                acting_user=acting_user,
+                modified_stream=sub.stream,
+                extra_data=orjson.dumps(
+                    {
+                        RealmAuditLog.OLD_VALUE: Subscription.ROLE_STREAM_ADMINISTRATOR,
+                        RealmAuditLog.NEW_VALUE: Subscription.ROLE_MEMBER,
```

Comment:
Can we add a "reason: "unsubscribe""` field in this block?  I imagine that detail may be important for future investigation.